feat: fetch recovery methods for current user

chadoh opened the pull request:
This adds some data-handling to make use of some of the data & endpoints added in nearprotocol/near-contract-helper#88

* refactor wallet to export instance, not class (see first commit for full explanation of & reasoning for this change)
* add action, reducer, and hook for fetching recovery methods
* fetch these on the profile page
* only fetch them if someone is viewing their own profile
* `addAccessKey` now POSTs to `near-contract-helper` to record the publicKey & time that this recovery method was added

## Testing

To make sure this works as expected, visit the [preview URL](https://near-wallet-staging-pr-399.onrender.com/) and create a new account. Keep your browser console open as you do this.

Visit your profile. In the Network tab of your inspector, you should see a successful POST to the `recoveryMethods` endpoint of `near-contract-helper`:

<img width="423" alt="Screen Shot 2020-03-30 at 09 56 39" src="https://user-images.githubusercontent.com/221614/77955729-afea7980-729e-11ea-9075-4c5c1673da82.png">

In the Console tab of your inspector, you should see the recovery methods logged out:

<img width="423" alt="chrome console showing recoveryMethods array of length one" src="https://user-images.githubusercontent.com/221614/78395344-4aa6c900-75bb-11ea-9ec7-6a007b0dcdd1.png">

And if you have the [Redux Dev Tools extension for Chrome](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) installed, you can visit the Redux tab of your inspector and you should be able to see recovery methods loaded in:

<img width="731" alt="Screen Shot 2020-04-03 at 14 58 37" src="https://user-images.githubusercontent.com/221614/78395626-ce60b580-75bb-11ea-92a7-5f3c310a971c.png">

vgrichina commented:
Is this hooked up? Seems to show fake data. Does this need to be combined with other PR from @Patrick1904?

<img width="1185" alt="Screen Shot 2020-03-10 at 12 17 35 AM" src="https://user-images.githubusercontent.com/126710/76288751-a99b4b80-6264-11ea-88c7-b27019ab0052.png">

chadoh commented:
@vgrichina this lays some groundwork for #381 – it does not use the data, it just fetches it for now. #381 can branch off of this, or wait until this one gets merged to `master`. I thought a standalone PR would be useful so that this logic could be vetted separately.

vgrichina commented on src/utils/wallet.js:
```diff
@@ -153,10 +154,31 @@ export class Wallet {
+            recoveryMethods.phone = { data: fromDb.phoneNumber, timeStamp: fromDb.phoneAddedAt }
+        }
+        if (fromDb.phraseAddedAt) {
+            recoveryMethods.phrase = { timeStamp: fromDb.phraseAddedAt }
```
@chadoh @Patrick1904  I think we have pretty big problem that none of these have info on what are associated public keys, so we have no idea which key to remove if you want to replace/remove given recovery method.

vgrichina replied:
My original plan to address it was to add `meta` to the access keys which is loaded from elsewhere:
https://github.com/nearprotocol/near-wallet/blob/b7bbfbeef97c64d668cfe07385b377a6d0bac6e3/src/utils/wallet.js#L170

chadoh replied:
Oh, great point! Should we be storing that information in the `near-contract-helper` database?

vgrichina replied:
@chadoh I think long term we should be storing this information on chain using open web contract from @evgenykuzyakov. We can store it in near-contract-helper short term if it's easier to get working.

chadoh replied:
Stored in near-contract-helper now (pending merge of nearprotocol/near-contract-helper#88)

vgrichina commented:
> this lays some groundwork for #381 – it does not use the data, it just fetches it for now. #381 can branch off of this, or wait until this one gets merged to master.

best to branch off this from #381 and get it connected. Stuff merged to `master` should be shippable (as it gets deployed as testnet wallet).

chadoh commented:
> Stuff merged to master should be shippable (as it gets deployed as testnet wallet).

This is shippable – it adds data to the redux store, and that's it. It has no adverse effect in the UI.

Patrick1904 commented:
> > Stuff merged to master should be shippable (as it gets deployed as testnet wallet).
> 
> This is shippable – it adds data to the redux store, and that's it. It has no adverse effect in the UI.

I agree. Once there are public keys associated with each recovery method, I think we should merge this to master. Once it's on master, I can hook it up with https://github.com/nearprotocol/near-wallet/pull/381 as there are some UI changes needed for it to work with this data structure.

vgrichina commented on src/utils/wallet.js:
```diff
@@ -153,10 +154,31 @@ export class Wallet {
+        ]);
+
+        const recoveryMethods = {}
+        if (fromDb.email) {
```
why not return data from contract helper in already desired format?

chadoh replied:
Done!

vgrichina commented:
> I agree. Once there are public keys associated with each recovery method, I think we should merge this to master. Once it's on master, I can hook it up with #381 as there are some UI changes needed for it to work with this data structure.

np, up to you folks to merge in whatever order works best, just make sure it doesn't bring anything down!

vgrichina commented:
@chadoh seems like it's working alright

> loadAccount now fetches recovery methods from near-contract-helper, and these recoveryMethods are added to the redux state in reducers/account/index

not sure why it's needed to load these every time we load account. Seems like it's only relevant for specific page (when recovery method setup is happening) unless I misunderstand smth.

vgrichina commented:
@Patrick1904 you might have better context here on when this info is needed (e.g. if we plan to use it to display banner on homepage to nudge user – then it indeed makes sense to have it in `loadAccount`) /cc @chadoh

Patrick1904 commented:
> @Patrick1904 you might have better context here on when this info is needed (e.g. if we plan to use it to display banner on homepage to nudge user – then it indeed makes sense to have it in `loadAccount`) /cc @chadoh

<img width="396" alt="Screen Shot 2020-02-06 at 11 14 40 AM" src="https://user-images.githubusercontent.com/24921205/78165741-a6792280-7400-11ea-8453-d7e7c0c0416d.png">

We'll need recovery information on the profile page for the recovery management module and I believe we're planning on using it to encourage setting up additional/specific methods as part of [communicating security expectations](https://github.com/nearprotocol/near-wallet/issues/442).

Patrick1904 commented:
![Screen Shot 2020-04-02 at 10 09 50 AM](https://user-images.githubusercontent.com/24921205/78277930-215a4000-74ca-11ea-84d5-91a5570a13aa.png)

@chadoh I'm not able to create account. Please see attached

Patrick1904 commented:
@chadoh, btw, not sure if this has anything to do with this PR, but recovery links sent using onrender don't get a domain resulting in: `http://undefined/recover-with-link/someusername/phrase`.

chadoh commented:
@Patrick1904 looks like we were missing the `WALLET_URL` environment variable in the betanet-account-helper service which gets used here: https://github.com/nearprotocol/near-contract-helper/blob/45645748e9879ab3865fd893604ce1c5b8147700/app.js#L219

I've updated `WALLET_URL` to point at https://near-wallet-pr-399.onrender.com, so you should see correct links in these emails now.

@vgrichina great point about this data not being needed app-wide, and it therefore being wasteful to fetch it in the main `account` reducer. I'll update the logic here to add a new action & reducer.

chadoh commented:
@Patrick1904 @vgrichina I've reworked the logic here and the PR description. I've also updated the environment variables for https://near-wallet-pr-399.onrender.com to point at the main testnet endpoints now that nearprotocol/near-contract-helper#88 has been merged. Please re-review and let me know what you think!

vgrichina commented on src/reducers/recoveryMethods.js:
```diff
@@ -0,0 +1,20 @@
+    if (!ready) return state
+
+    if (error) {
+        console.error('error loading recoveryMethods!', payload)
```
nit: this is redundant, should be enough to have logging once in the middleware for executing promises

I would just do `if (!ready || error) return state` above (unless there is need to update state based on error).

vgrichina commented on src/utils/wallet.js:
```diff
@@ -254,12 +250,20 @@ export class Wallet {
             ACCESS_KEY_FUNDING_AMOUNT
         )
+
+        await sendJson('POST', `${ACCOUNT_HELPER_URL}/account/seedPhraseAdded`, {
```
this only needs to happen for `addAccessKeySeedPhrase` action, having it here would result in every app-specific access key to be registered as a seed phrase

chadoh replied:
Great point! I've moved this logic to the action creator – I don't love the look of it, but maybe it's ok. Do you see a quick way to clean it up?

I have verified that it still works locally:

<img width="731" alt="Screen Shot 2020-04-07 at 13 02 33" src="https://user-images.githubusercontent.com/221614/78698427-51da1980-78d0-11ea-9fc2-e543a0e48916.png">

vgrichina replied:
@chadoh I think it makes sense to move more logic to actions. `wallet.js` seems like just historically ended up being almost "god object" kind of thing.

We probably just need to modularize actions better (for now way too much stuff ended up in `account.js`)
